File: src/main/java/com/lessmarkup/engine/module/ModuleProviderImpl.java
package com.lessmarkup.engine.module;

import com.lessmarkup.dataobjects.Module;
import com.lessmarkup.framework.helpers.DependencyResolver;
import com.lessmarkup.framework.helpers.LoggingHelper;
import com.lessmarkup.framework.system.RequestContextHolder;
import com.lessmarkup.interfaces.data.DomainModel;
import com.lessmarkup.interfaces.data.DomainModelProvider;
import com.lessmarkup.interfaces.module.ModuleConfiguration;
import com.lessmarkup.interfaces.module.ModuleInitializer;
import com.lessmarkup.interfaces.module.ModuleProvider;
import com.lessmarkup.interfaces.structure.NodeHandler;
import com.lessmarkup.interfaces.structure.Tuple;
import org.springframework.stereotype.Component;

import java.io.*;
import java.lang.reflect.Modifier;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLClassLoader;
import java.util.*;
import java.util.stream.Collectors;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

@Component
public class ModuleProviderImpl implements ModuleProvider {

    private final List<ModuleConfiguration> modules = new ArrayList<>();
    private final Map<String, Tuple<Class<? extends NodeHandler>, String>> nodeHandlers = new HashMap<>();

    @Override
    public Collection<ModuleConfiguration> getModules() {
        return modules;
    }

    @Override
    public void discoverAndRegisterModules() {
        
        URLClassLoader classLoader = (URLClassLoader) getClass().getClassLoader();

        for (URL url : classLoader.getURLs()) {
            if (!Objects.equals(url.getProtocol(), "file")) {
                continue;
            }
            if (url.getPath().endsWith(".jar")) {
                // we only load from file system to ignore dependent JARs
                continue;
            }
            discoverModules(url, true, classLoader);
        }

        String modulesPath = RequestContextHolder.getContext().getEngineConfiguration().getModulesPath();
        
        if (modulesPath == null || modulesPath.length() == 0) {
            return;
        }
        
        File modulesDirectory = new File(modulesPath);

        if (modulesDirectory.exists()) {
            for (File file : modulesDirectory.listFiles()) {
                if (!file.getName().endsWith(".jar")) {
                    continue;
                }
                try {
                    discoverModules(file.toURI().toURL(), false, null);
                } catch (MalformedURLException ex) {
                    LoggingHelper.logException(getClass(), ex);
                }
            }
        }
    }
    
    private void listAllFileSystemModuleElements(String basePath, File directory, List<String> elements) {

        File[] files = directory.listFiles();
        if (files == null) {
            String userName = System.getProperty("user.name");
            try (PrintWriter writer = new PrintWriter(System.out)) {
                writer.write(userName);
            }
            return;
        }

        for (File file : files) {
            if (file.isDirectory()) {
                listAllFileSystemModuleElements(basePath, file, elements);
            } else {
                String filePath;
                try {
                    filePath = file.toURI().toURL().getPath();
                    if (filePath.startsWith(basePath)) {
                        elements.add(filePath.substring(basePath.length()));
                    }
                } catch (MalformedURLException ex) {
                    LoggingHelper.logException(getClass(), ex);
                }
            }
        }
    }

    private List<String> listAllFileSystemModuleElements(String path) {
        List<String> ret = new LinkedList<>();
        File moduleDirectory = new File(path);
        try {
            listAllFileSystemModuleElements(moduleDirectory.toURI().toURL().getPath(), moduleDirectory, ret);
        } catch (MalformedURLException ex) {
            LoggingHelper.logException(getClass(), ex);
        }
        return ret;
    }

    private List<String> listAllJarModuleElements(String path) {
        try (InputStream stream = new FileInputStream(path); ZipInputStream zip = new ZipInputStream(stream)) {
            List<String> ret = new LinkedList<>();
            for (;;) {
                ZipEntry entry = zip.getNextEntry();
                if (entry == null) {
                    break;
                }

                ret.add(entry.getName());
            }

            return ret;
        } catch (IOException ex) {
            LoggingHelper.logException(getClass(), ex);
            return null;
        }
    }

    private List<String> listAllModuleElements(String path) {
        if (path.endsWith(".jar")) {
            return listAllJarModuleElements(path);
        }

        return listAllFileSystemModuleElements(path);
    }

    private void constructModule(URL moduleUrl, boolean isSystem, String classPath, ClassLoader classLoader, List<String> elements) {

        if (!classPath.endsWith(".class")) {
            return;
        }
        
        ModuleInitializer moduleInitializer;

        try {
            String className = classPath.substring(0, classPath.length()-".class".length()).replaceAll("/", ".");
            Class<?> type = Class.forName(className, true, classLoader);
            if (type.isInterface() || Modifier.isAbstract(type.getModifiers()) || !ModuleInitializer.class.isAssignableFrom(type)) {
                return;
            }
            moduleInitializer = (ModuleInitializer) DependencyResolver.resolve(type);
            if (moduleInitializer == null) {
                return;
            }
        } catch (SecurityException | ClassNotFoundException | IllegalArgumentException ex) {
            LoggingHelper.logException(getClass(), ex);
            return;
        }

        moduleInitializer.initialize();

        ModuleConfigurationImpl moduleConfiguration = new ModuleConfigurationImpl(moduleUrl, isSystem, moduleInitializer.getModuleType(), elements, classLoader, moduleInitializer);

        modules.add(moduleConfiguration);

        for (Class<? extends NodeHandler> nodeHandler : moduleInitializer.getNodeHandlerTypes()) {
            nodeHandlers.put(nodeHandler.getSimpleName(), new Tuple<>(nodeHandler, moduleConfiguration.getModuleType()));
        }
    }
    
    private void discoverModules(URL moduleUrl, boolean isSystem, ClassLoader classLoader) {

        if (!moduleUrl.getProtocol().equals("file")) {
            return;
        }

        List<String> elements = listAllModuleElements(moduleUrl.getPath());
        List<String> moduleInitializerClasses = elements.stream()
                .filter(path -> path.endsWith("ModuleInitializer.class"))
                .collect(Collectors.toCollection(LinkedList::new));

        if (moduleInitializerClasses.isEmpty()) {
            LoggingHelper.getLogger(getClass()).info(String.format("Cannot find initalizer for module %s", moduleUrl));
            return;
        }

        if (classLoader == null) {
            classLoader = URLClassLoader.newInstance(new URL[]{moduleUrl}, getClass().getClassLoader());
        }

        for (String initializerClass : moduleInitializerClasses) {
            constructModule(moduleUrl, isSystem, initializerClass, classLoader, elements);
        }
    }

    @Override
    public void updateModuleDatabase(DomainModelProvider domainModelProvider) {
        if (domainModelProvider == null) {
            return;
        }
        
        try (DomainModel domainModel = domainModelProvider.create()) {
            List<ModuleConfiguration> existingModules = new ArrayList<>();
            domainModel.query().from(Module.class).where("removed = $", false).toList(Module.class).forEach(module -> {
                Optional<ModuleConfiguration> reference = modules.stream().filter(m -> m.getUrl().toString().equals(module.getPath())).findFirst();
                if (!reference.isPresent()) {
                    module.setRemoved(true);
                } else {
                    existingModules.add(reference.get());
                    module.setSystem(reference.get().isSystem());
                    module.setModuleType(reference.get().getModuleType());
                }
                domainModel.update(module);
            });
            
            modules.stream().filter(m -> !existingModules.contains(m)).forEach(configuration -> {
                Module module = new Module();
                module.setEnabled(true);
                module.setName(configuration.getInitializer().getName());
                module.setPath(configuration.getUrl().toString());
                module.setRemoved(false);
                module.setSystem(configuration.isSystem());
                module.setModuleType(configuration.getModuleType());
                domainModel.create(module);
            });
        }
    }

    @Override
    public Collection<String> getNodeHandlers() {
        return nodeHandlers.keySet();
    }

    @Override
    public Tuple<Class<? extends NodeHandler>, String> getNodeHandler(String id) {
        return nodeHandlers.get(id);
    }
}

File: src/main/scala/com/lessmarkup/engine/module/ModuleProviderImpl.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.engine.module

import java.io._
import java.lang.reflect.Modifier
import java.net.{MalformedURLException, URL, URLClassLoader, URLDecoder}
import java.util.zip.ZipInputStream
import com.google.inject.Singleton
import com.lessmarkup.dataobjects.Module
import com.lessmarkup.framework.helpers.{DependencyResolver, LoggingHelper}
import com.lessmarkup.framework.system.RequestContextHolder
import com.lessmarkup.interfaces.annotations.Implements
import com.lessmarkup.interfaces.data.{DomainModel, DomainModelProvider}
import com.lessmarkup.interfaces.exceptions.CommonException
import com.lessmarkup.interfaces.module.{ModuleConfiguration, ModuleInitializer, ModuleProvider}
import com.lessmarkup.interfaces.structure.NodeHandlerFactory
import scala.util.Try

class ModuleProviderImpl(onlySystem: Boolean) extends ModuleProvider {

  private val modules = discoverAndRegisterModules()
  private val nodeHandlers: Map[String, (Class[_ <: NodeHandlerFactory], String)] = modules
    .flatMap(m => m.getInitializer.getNodeHandlerTypes.map(h => (m, h)))
    .map(h => (h._2.getSimpleName, (h._2, h._1.getModuleType))).toMap

  def this() = {
    this(false)
  }

  def getModules: Seq[ModuleConfiguration] = modules

  def updateModuleDatabase(domainModelProvider: DomainModelProvider) {

    if (domainModelProvider == null) {
      return
    }

    val domainModel: DomainModel = domainModelProvider.create
    try {

      val query = domainModel.query.from(classOf[Module]).where("removed = $", new java.lang.Boolean(false))
      val databaseModules = query.toList(classOf[Module])
      val existingModules = modules.map(m => m.getUrl.toString -> m).toMap

      databaseModules.filter(m => !existingModules.contains(m.path)).foreach(m => {
        m.removed = true
        domainModel.update(m)
      })

      val existingDatabaseModules = databaseModules.filter(m => existingModules.contains(m.path))

      existingDatabaseModules.foreach(m => {
        val existingModule = existingModules.get(m.path).get
        m.system = existingModule.isSystem
        m.moduleType = existingModule.getModuleType
        domainModel.update(m)
      })

      val existingDatabasePaths = databaseModules.map(_.path).toSet

      modules.filter(m => !existingDatabasePaths.contains(m.getUrl.toString)).foreach(m => {
        val module = new Module
        module.enabled = true
        module.name = m.getInitializer.getName
        module.path = m.getUrl.toString
        module.removed = false
        module.system = m.isSystem
        module.moduleType = m.getModuleType
        domainModel.create(module)
      })

    } finally {
      if (domainModel != null) domainModel.close()
    }
  }

  def getNodeHandlers: Seq[String] = nodeHandlers.keySet.toSeq

  def getNodeHandler(id: String): Option[(Class[_ <: NodeHandlerFactory], String)] = {
    nodeHandlers.get(id)
  }

  private def discoverAndRegisterModules(): Seq[ModuleConfiguration] = {

    DependencyResolver.reset(this)

    val classLoader: URLClassLoader = getClass.getClassLoader.asInstanceOf[URLClassLoader]

    val systemModules = (for (url <- classLoader.getURLs if url.getProtocol == "file" if !url.getPath.endsWith(".jar"))
        yield discoverModules(url, isSystem = true, Option(classLoader))
      ).flatten.toList

    if (onlySystem) {
      return systemModules
    }

    val modulesPath: String = RequestContextHolder.getContext.getEngineConfiguration.getModulesPath
    if (modulesPath == null || modulesPath.length == 0) {
      return systemModules
    }

    val modulesDirectory: File = new File(modulesPath)
    if (!modulesDirectory.exists) {
      return systemModules
    }

    systemModules ::: (for (file <- modulesDirectory.listFiles if file.getName.endsWith(".jar"))
      yield discoverModules(file.toURI.toURL, isSystem = false, null)).flatten.toList
  }

  private def directoryEntryPath(basePath: String, node: File) = {
    val fullPath = node.toURI.toURL.getPath
    if (fullPath.startsWith(basePath))
      fullPath.substring(basePath.length)
    else
      fullPath
  }

  private def directoryGetFiles(node: File): Array[File] = {

    if (node.isFile) {
      return Array()
    }

    val files = node.listFiles
    if (files == null) {
      val userName: String = System.getProperty("user.name")
      val writer: PrintWriter = new PrintWriter(System.out)

      try {
        writer.write(userName)
      } finally {
        if (writer != null) writer.close()
      }

      throw new CommonException("Access denied for user: " + userName + " and file:" + node.getPath)
    }
    files
  }

  private def flattenDirectoryTree(basePath: String, nodes: Array[File]) : List[String] =
    if (nodes.isEmpty) Nil
    else if (nodes.head.isDirectory) flattenDirectoryTree(basePath, directoryGetFiles(nodes.head)) ::: flattenDirectoryTree(basePath, nodes.tail)
    else directoryEntryPath(basePath, nodes.head) :: flattenDirectoryTree(basePath, nodes.tail)

  private def listAllFileSystemModuleElements(basePath: String, directory: File) : List[String] = {
    flattenDirectoryTree(basePath, directoryGetFiles(directory))
  }

  private def listAllFileSystemModuleElements(path: String): List[String] = {
    val moduleDirectory: File = new File(path)
    try {
      listAllFileSystemModuleElements(moduleDirectory.toURI.toURL.getPath, moduleDirectory)
    }
    catch {
      case ex:
        MalformedURLException => LoggingHelper.logException(getClass, ex)
        List[String]()
    }
  }

  private def listAllJarModuleElements(path: String) = {
    val stream: InputStream = new FileInputStream(path)
    val zip: ZipInputStream = new ZipInputStream(stream)
    try {
      Stream.continually(zip.getNextEntry).takeWhile(_ != null).map(_.getName).toList
    }
    catch {
      case ex: IOException =>
        LoggingHelper.logException(getClass, ex)
        List[String]()
    } finally {
      if (stream != null)
        stream.close()
      if (zip != null)
        zip.close()
    }
  }

  private def listAllModuleElements(path: String) = {
    if (path.endsWith(".jar")) {
      listAllJarModuleElements(path)
    } else {
      listAllFileSystemModuleElements(path)
    }
  }

  private def constructModule(moduleUrl: URL, isSystem: Boolean, moduleInitializerClass: Class[_ <: AnyRef], classLoader: ClassLoader, elements: List[String]): Option[ModuleConfiguration] = {

    if (moduleInitializerClass.isInterface || Modifier.isAbstract(moduleInitializerClass.getModifiers) || !classOf[ModuleInitializer].isAssignableFrom(moduleInitializerClass)) {
      return None
    }

    val moduleInitializer = DependencyResolver(moduleInitializerClass).asInstanceOf[ModuleInitializer]

    if (moduleInitializer == null) {
      return None
    }

    moduleInitializer.initialize()

    val moduleConfiguration = new ModuleConfigurationImpl(moduleUrl, isSystem, moduleInitializer.getModuleType, elements, classLoader, moduleInitializer)

    Option(moduleConfiguration)
  }

  private def discoverModules(moduleUrl: URL, isSystem: Boolean, classLoader: Option[ClassLoader]): List[ModuleConfiguration] = {

    if (!(moduleUrl.getProtocol == "file")) {
      return List[ModuleConfiguration]()
    }

    val elements = Try({
      listAllModuleElements(URLDecoder.decode(moduleUrl.getPath, "UTF-8"))
    })
      .recoverWith { case ex: Throwable => ex.printStackTrace(); return List[ModuleConfiguration]() }
      .getOrElse(List[String]())

    val initializedClassLoader = classLoader.getOrElse(URLClassLoader.newInstance(Array[URL](moduleUrl), getClass.getClassLoader))

    val classes: List[Class[_ <: AnyRef]] =
      elements
        .filter(_.endsWith(".class"))
        .map(path => path.substring(0, path.length - ".class".length).replaceAll("/", "."))
        .filterNot(_.contains("$"))
        .flatMap(
          className => {
            val cl: Option[Class[_ <: AnyRef]] = try {
              val loadedClass = Class.forName(className, true, initializedClassLoader)
              Option(loadedClass.asInstanceOf[Class[_ <: AnyRef]])
            } catch {
              case _: Throwable =>
                None
            }
            cl
          })


    val moduleInitializerClasses = classes.filter(c => classOf[ModuleInitializer].isAssignableFrom(c) && !Modifier.isAbstract(c.getModifiers) && !Modifier.isInterface(c.getModifiers))

    if (moduleInitializerClasses.isEmpty) {
      return Nil
    }

    DependencyResolver.add(classes)

    moduleInitializerClasses
      .map(constructModule(moduleUrl, isSystem, _, initializedClassLoader, elements))
      .filter(_.isDefined)
      .map(_.get)
  }
}